### DISCUSSION TITLE
Tacotron 2 with parallelwavegan next step

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#263] [Archived]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 28, 2020, 5:39am]<br/> <br/> Hi<br/> <br/> So i trained Mozilla TTS with Tacotron2 using a custom dataset. The<br/> Griffin Lim previews are starting to sound really good, although<br/> robotic.<br/> <br/> I now want to move on to use the ParallelWaveGAN vocoder.<br/> <br/> 1. How do i go about doing that? Is there a notebook or syntax to run<br/> it?<br/> 2. What are the prerequisits?<br/> 3. Do i need to train the ParallelWaveGAN model at all, like i did with<br/> the TTS model?<br/> <br/> I just want to understand the full process before starting to refine my<br/> original dataset and possibly retrain my tts model with larger set and<br/> better quality audio.<br/> <br/> I am stuck at the TTS training stage right now.<br/> <br/> Any pointers would be greatly appreciated<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/tacotron-2-with-parallelwavegan-next-step]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> dkreutz<br/> <DATE> [August 28, 2020, 3:12pm]<br/> <br/> You need to run the „extract features' notebook, then start the PwGAN<br/> training. Make sure to use the same settings for audio processing<br/> (sample rate, trim silence, normalize, fmin, fmax, etc.) as you have set<br/> for the Taco2 training.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 6:40am]<br/> <br/> Thanks but I couldnt find this *slash* 'extract features' notebook. In my TTS<br/> installation there are a couple of them and one is called<br/> 'ExtractTTSpectrogram.ipynb'.<br/> <br/> However the description says *slash* 'This is a notebook to generate<br/> mel-spectrograms from a TTS model to be used for WaveRNN training.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> dkreutz<br/> <DATE> [August 29, 2020, 8:07am]<br/> <br/> Sorry, I mixed this up. For the PwGAN you need to run a preprocess<br/> script (which basically does the same as the extract-features notebook<br/> for WaveRNN).<br/> <br/> Which PwGAN implementation are you going to use?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> sanjaesc<br/> <DATE> [August 29, 2020, 8:11am]<br/> <br/> What branch are you using?<br/> <br/> You can also train on ground truth, just set the 'data_path': in the<br/> config to the path where all your wav's are. *slash* <br/> Set the audio parameters equal your TTS settings.<br/> <br/> Start the training as stated here:<br/> <https://github.com/mozilla/TTS/tree/dev/mozilla_voice_tts/vocoder><br/> <br/> You can also check this <https://github.com/erogol/TTS_recipes><br/> repository, which contains full recipes for training TTS and Vocoders.<br/> You can try adjusting them to your needs.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:21am]<br/> <br/> > ipes for training TTS and Vocoders. You can try adjusting them to<br/> <br/> Hi<br/> <br/> I am using the Mozilla TTS dev branch<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:28am]<br/> <br/> Sorry for being a stupid noob. I am not sure what you mean by that.<br/> <br/> I am using Colab for the TTS and was planning to use this<br/> <br/> ![](https://github.githubassets.com/favicons/favicon.svg){.site-icon<br/> [GitHub](https://github.com/erogol/ParallelWaveGAN)<br/> <br/> ### [erogol/ParallelWaveGAN](https://github.com/erogol/ParallelWaveGAN)<br/> <br/> ParallelWaveGAN adaptation for Mozilla TTS. Contribute to<br/> erogol/ParallelWaveGAN development by creating an account on GitHub.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:32am]<br/> <br/> Honestly, i dont care which vocoder i use. I just want the best sounding<br/> one. I have no need for real time TTS, so i just want to render<br/> sentences to audio 'offline' so to speak. Do you have any<br/> recommendations based on that?<br/> <br/> The TTS model is trained at this point, so now i just want it to sound<br/> better.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> sanjaesc<br/> <DATE> [August 29, 2020, 8:34am]<br/> <br/> This is the old implementation.<br/> <br/> The current dev branch has a nativ vocoder support:<br/> <https://github.com/mozilla/TTS/tree/dev/mozilla_voice_tts/vocoder><br/> <br/> Under the configs folder check the<br/> [parallel_wavegan_config.json](https://github.com/mozilla/TTS/blob/dev/mozilla_voice_tts/vocoder/configs/parallel_wavegan_config.json).<br/> <br/> Follow my steps from above: *slash* <br/> 1: Set [audio<br/> parameters](https://github.com/mozilla/TTS/blob/34241815e7025b30dddb03196e113ced3e4f2906/mozilla_voice_tts/vocoder/configs/parallel_wavegan_config.json#L6-L34) *slash* <br/> 2: Set [path to your wav<br/> files](https://github.com/mozilla/TTS/blob/34241815e7025b30dddb03196e113ced3e4f2906/mozilla_voice_tts/vocoder/configs/parallel_wavegan_config.json#L92) *slash* <br/> 3: [Start<br/> training](https://github.com/mozilla/TTS/blob/34241815e7025b30dddb03196e113ced3e4f2906/mozilla_voice_tts/vocoder/configs/parallel_wavegan_config.json#L6-L34)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:35am]<br/> <br/> Ok, so no extraction step needed?<br/> <br/> I was hoping to use a pretrained vocoder model to reduce training time.<br/> So i need to start from scratch based on my trained TTS model?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:36am]<br/> <br/> > The current dev branch has a nativ vocoder support:<br/> > <https://github.com/mozilla/TTS/tree/dev/mozilla_voice_tts/vocoder><br/> <br/> I am cloning this into my colab notebook at the moment *slash* <br/> !git clone -b dev <https://github.com/mozilla/TTS><br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:42am]<br/> <br/> > Using CUDA: True *slash* <br/> > Number of GPUs: 1 *slash* <br/> > usage: train_vocoder.py *slash* [-h *slash* ] *slash* --continue_path CONTINUE_PATH *slash* <br/> > *slash* [--restore_path RESTORE_PATH *slash* ] *slash* --config_path *slash* <br/> > CONFIG_PATH *slash* [--debug DEBUG *slash* ] *slash* [--rank RANK *slash* ] *slash* <br/> > *slash* [--group_id GROUP_ID *slash* ] *slash* <br/> > train_vocoder.py: error: the following arguments are required:<br/> > *slash* --continue_path, *slash* --config_path<br/> <br/> Are these paths to the TTS model and TTS config file or the Vocoder<br/> ones?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:45am]<br/> <br/> Thanks for your help!!<br/> <br/> I tried this<br/> <br/> !python /content/TTS/mozilla_voice_tts/bin/train_vocoder.py<br/> *slash* --config_path<br/> '/content/TTS/mozilla_voice_tts/vocoder/configs/parallel_wavegan_config.json'<br/> <br/> It seems to be working.<br/> <br/> 1. Where is the vocoder model saved then? I configured my TTS model to<br/> be saved direcly to Drive, and it would be cool if I could do the<br/> same for the Vocoder<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> sanjaesc<br/> <DATE> [August 29, 2020, 8:48am]<br/> <br/> > Ok, so no extraction step needed?<br/> <br/> Yes, works also without extraction. There is also an option to train on<br/> extracted features but I haven't tested it yes.<br/> <br/> > Are these paths to the TTS model and TTS config file or the Vocoder<br/> > ones?<br/> <br/> Those are paths to the vocoder, if you want to fine_tune or continue<br/> training later on.<br/> <br/> > Where is the vocoder model saved then? I configured my TTS model to be<br/> > saved direcly to Drive, and it would be cool if I could do the same<br/> > for the Vocoder<br/> <br/> In the config you can set the output_path, somewhere at the bottom.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 8:51am]<br/> <br/> Thanks man, i REALLY appreciate it!<br/> <br/> 1. So the Vocoder and the TTS model are completly separate although<br/> trained on the same data set?<br/> <br/> 2. When i later have trained the vocoder on the same dataset, these two<br/> are combined/working together to synthesize the sentence? Is this a<br/> correct description?<br/> <br/> 3. At what point will i be able to test these models together? Is there<br/> a notebook available that allows me to point to my TTS and Vocoder<br/> model to generate audio?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> dkreutz<br/> <DATE> [August 29, 2020, 9:32am]<br/> <br/> 1. yes.<br/> <br/> 2. In a very generalized way: TTS models are trained on audio+text<br/> input and output mel-spectrograms. Vocoder models are trained on<br/> mel-specs input and has audio output. *slash* <br/> (Griffin-Lim vocoder is a special case - it interpolates directly<br/> audio from mel-specs without a model, but quality is inferior<br/> compared to vocoder models)<br/> <br/> 3. synthesize.py in dev-branch has vocoder-path option where you can<br/> specify vocoder model. I believe there is also a notebook...<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 9:32am]<br/> <br/> Thanks a lot! This really helps me to learn<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 12:12pm]<br/> <br/> > In a very generalized way: TTS models are trained on audio+text input<br/> > and output mel-spectrograms. Vocoder models are trained on mel-specs<br/> > input and has audio output. *slash* <br/> > (Griffin-Lim vocoder is a special case - it interpolates directly<br/> > audio from mel-specs without a model, but quality is inferior compared<br/> > to vocoder models)<br/> <br/> But you NEED to train the TTS model before training the vocoder, right?<br/> Cant have one without the other?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> dkreutz<br/> <DATE> [August 29, 2020, 12:16pm]<br/> <br/> You can train the vocoder first, but it is of no use standalone as it<br/> needs mel-specs from the TTS as input for inference (synthesis).<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> CrazyJoeDevola<br/> <DATE> [August 29, 2020, 12:21pm]<br/> <br/> Wouldnt it make sense to use a pretrained vocoder model instead of<br/> starting from scratch like i did?<br/> <br/> [**[next page<br/> →](https://discourse.mozilla.org/t/tacotron-2-with-parallelwavegan-next-step/66546?page=2)**<br/> <br/> [Archived Post]<br/>
